qa/workunits/rbd/luks-encryption.sh

```diff
@@ -34,24 +35,109 @@ function test_encryption_format() {
+  clean_up_cryptsetup
+
+  # write 1MB plaintext
+  dd if=/tmp/testdata1 of=$RAW_DEV oflag=direct bs=1M count=1
+
+  # clone (luks1)
+  rbd snap create testimg@snap
+  rbd snap protect testimg@snap
+  rbd clone testimg@snap testimg1
+  rbd encryption format testimg1 luks1 /tmp/passphrase
+
+  # open encryption with librbd, write one more MB, close
+  LIBRBD_DEV=$(_sudo rbd -p rbd map testimg1 -t nbd -o encryption-format=luks1,encryption-passphrase-file=/tmp/passphrase)
+  sudo chmod 666 $LIBRBD_DEV
+  dd if=$LIBRBD_DEV of=/tmp/cmpdata iflag=direct bs=1M count=1
+  cmp -n 1MB /tmp/cmpdata /tmp/testdata1
+  dd if=/tmp/testdata1 of=$LIBRBD_DEV seek=1 skip=1 oflag=direct bs=1M count=1
+  sudo rbd device unmap -t nbd $LIBRBD_DEV
+
+  # second clone (luks2)
+  rbd snap create testimg1@snap
+  rbd snap protect testimg1@snap
+  rbd clone testimg1@snap testimg2
+  rbd encryption format testimg2 luks2 /tmp/passphrase2
+
+  # open encryption with librbd, write one more MB, close
+  LIBRBD_DEV=$(_sudo rbd -p rbd map testimg2 -t nbd -o encryption-format=luks2,encryption-passphrase-file=/tmp/passphrase2,encryption-format=luks1,encryption-passphrase-file=/tmp/passphrase)
+  sudo chmod 666 $LIBRBD_DEV
+  dd if=$LIBRBD_DEV of=/tmp/cmpdata iflag=direct bs=1M count=2
+  cmp -n 2MB /tmp/cmpdata /tmp/testdata1
+  dd if=/tmp/testdata1 of=$LIBRBD_DEV seek=2 skip=2 oflag=direct bs=1M count=1
+  sudo rbd device unmap -t nbd $LIBRBD_DEV
+
+  # flatten
+  rbd flatten testimg2 --encryption-format luks --encryption-format luks --encryption-passphrase-file /tmp/passphrase2 --encryption-passphrase-file /tmp/passphrase
+
+  # verify with cryptsetup
+  RAW_FLAT_DEV=$(_sudo rbd -p rbd map testimg2 -t nbd)
+  sudo cryptsetup open $RAW_FLAT_DEV --type luks cryptsetupdev -d /tmp/passphrase2
+  sudo chmod 666 /dev/mapper/cryptsetupdev
+  dd if=/dev/mapper/cryptsetupdev of=/tmp/cmpdata iflag=direct bs=1M count=3
+  cmp -n 3MB /tmp/cmpdata /tmp/testdata1
+  sudo rbd device unmap -t nbd $RAW_FLAT_DEV
+}
+
+function test_clone_and_load_with_a_single_passphrase {
+  local expectedfail=$1
+
+  # clone and format
+  rbd snap create testimg@snap
+  rbd snap protect testimg@snap
+  rbd clone testimg@snap testimg1
+  rbd encryption format testimg1 luks2 /tmp/passphrase2
+
+  if [ "$expectedfail" = "true" ]
+  then
+    rbd flatten testimg1 --encryption-format luks --encryption-passphrase-file /tmp/passphrase2 && exit 1 || true
+  else
+    rbd flatten testimg1 --encryption-format luks --encryption-passphrase-file /tmp/passphrase2
+  fi
+
+  rbd remove testimg1 || true
+  rbd snap unprotect testimg@snap || true
+  rbd snap remove testimg@snap || true
+}
+
+function test_plaintext_detection {
+  # 16k LUKS header
+  sudo cryptsetup -q luksFormat --type luks2 --luks2-metadata-size 16k $RAW_DEV /tmp/passphrase
+  test_clone_and_load_with_a_single_passphrase true
+
+  # 4m LUKS header
+  sudo cryptsetup -q luksFormat --type luks2 --luks2-metadata-size 4m $RAW_DEV /tmp/passphrase
```

Comment:
Sorry I'm a bit confused.

First read is always 288KB.
Next reads (up to 4MB, and upto keyslots ends offset) are determined by the return values of libcryptsetup load and read_volume_key, respectively.

I'm not sure what arguments I should give to cryptsetup (e.g. `luks2-metadata-size`, `luks2-keyslots-size`, add dummy keyslots to push back "our" keyslot to a further offset) to satisfy the 3 scenarios you mentioned above.
I'm actually not sure that all scenarios can be crafted via cryptsetup.

Reply:
> First read is always 288KB. Next reads (up to 4MB, and upto keyslots ends offset) are determined by the return values of libcryptsetup load and read_volume_key, respectively.

Right, but note that there is another "next read" trigger for clones on the `replace_magic` error path.

> 
> I'm not sure what arguments I should give to cryptsetup (e.g. `luks2-metadata-size`, `luks2-keyslots-size`, add dummy keyslots to push back "our" keyslot to a further offset) to satisfy the 3 scenarios you mentioned above. I'm actually not sure that all scenarios can be crafted via cryptsetup.

... and yet there is code that is supposed to handle these return values/scenarios.  Are you saying that it was added despite the lack of a way to trigger it?  If so, I'm concerned because some of this code was broken and it took multiple attempts to get it into "no obvious bugs" shape.  Was it actually tested?

Reply:
I mentioned dummy keyslots mostly in passing -- I assumed you already had all these cryptsetup invocations on hand from past experimentation.  I'll dig up the LUKS spec and see if I can come up with something concrete when I get a moment.

Reply:
> ... and yet there is code that is supposed to handle these return values/scenarios. Are you saying that it was added despite the lack of a way to trigger it? If so, I'm concerned because some of this code was broken and it took multiple attempts to get it into "no obvious bugs" shape. Was it actually tested?

Well, libcryptsetup assumes access to an image using a file.
We create a memory-mapped file, but obviously we don't want to read the entire image and write it to the mapped file.
So, we "guess" the right amount, and then yield to libcryptsetup.
In case libcryptsetup, we retry using the call to libcryptsetup, but this time feeding it with enough bytes (the maximum number, according to the LUKS spec).

So, our re-reads are not based on actual cases we know that exist, but rather are defensive actions that we take to ensure libcryptsetup didn't error due to lack of data.

To sum up, the actual liveness of the possible flows in the code depends on the internals of libcryptsetup, which I didn't bother to explore, as they also are not guaranteed to stay the same as new versions are out.

I tried my best to test this flows in the unit tests:
`TEST_F(TestMockCryptoLuksLoadRequest, KeyslotsBiggerThanInitialRead)`
`TEST_F(TestMockCryptoLuksLoadRequest, HeaderBiggerThanInitialRead)`

I may have missed some flows.
But anyways, I think it will be easier to mock libcryptsetup (by mocking Header.cc), than relaying on libcryptsetup internal implementation logic.

Reply:
Yeah, I took a look at the LUKS spec and there appears to be a provision for passphrase tokens with arbitrary (read: large) payloads but it can't be done with `cryptsetup` and is probably too much hassle anyway.  Could you please please re-add `--luks2-metadata-size 4m` case like you had before, just so some variability is still there even if it doesn't actually trigger much?  Apologies for this obstruction...